### DISCUSSION TITLE
在 umi max 中配置 routes 时如何使用自定义 icon？

### DISCUSSION THREAD

**Discussion by mist0616** [umijs/umi#10277] [Q&A]
在 umi max 中配置 routes 时如何使用自定义 icon？

**Comment by mist0616** [comment]
刚刚自测运行时 layout 配置 menuDataRender 可以解决？但好麻烦，请问有更好的解决方案吗？

**Comment by fz6m** [comment]
参考 layout 对 icon 支持的实现 `src/.umi/plugin-layout/runtime.tsx` ：

```ts
export function patchRoutes({ routes }) {
  Object.keys(routes).forEach(key => {
    const { icon } = routes[key];
    if (icon && typeof icon === 'string') {
      const upperIcon = formatIcon(icon);
      if (icons[upperIcon] || icons[upperIcon + 'Outlined']) {
        routes[key].icon = React.createElement(icons[upperIcon] || icons[upperIcon + 'Outlined']);
      }
    }
  });
}
```

自己在 `src/app.tsx` 里用 [`patchRoutes`](https://umijs.org/docs/api/runtime-config#patchroutes-routes-) 仿照着 patch 一下，把 icon 改成 react component

**Reply by mist0616** [reply]
> 参考 layout 对 icon 支持的实现 `src/.umi/plugin-layout/runtime.tsx` ：
> 
> ```ts
> export function patchRoutes({ routes }) {
>   Object.keys(routes).forEach(key => {
>     const { icon } = routes[key];
>     if (icon && typeof icon === 'string') {
>       const upperIcon = formatIcon(icon);
>       if (icons[upperIcon] || icons[upperIcon + 'Outlined']) {
>         routes[key].icon = React.createElement(icons[upperIcon] || icons[upperIcon + 'Outlined']);
>       }
>     }
>   });
> }
> ```
> 
> 自己在 `src/app.tsx` 里用 [`patchRoutes`](https://umijs.org/docs/api/runtime-config#patchroutes-routes-) 仿照着 patch 一下，把 icon 改成 react component

好的，感谢

**Comment by Eliot00** [comment]
可以用icons功能了 https://github.com/umijs/umi/pull/13396
